src/libsyntax/ext/expand.rs

```diff
@@ -278,13 +266,48 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
             self.cx.current_expansion = invoc.expansion_data.clone();
 
             self.cx.current_expansion.mark = scope;
-            let expansion = match ext {
-                Some(ext) => self.expand_invoc(invoc, ext),
-                None => invoc.expansion_kind.dummy(invoc.span()),
+            let (expansion, new_invocations) = if let Some(ext) = ext {
```

Comment:
This seems a bit complex, should there be a separate expand_derives function? (Two `Some(ext)`'s?)